internal/webhooks/patch_validation.go

```diff
@@ -0,0 +1,333 @@
+			return allErrs
+		}
+
+		if jsonPatch.Value == nil && jsonPatch.ValueFrom == nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch must define one of value or valueFrom.",
+				))
+		}
+
+		if jsonPatch.Value != nil && jsonPatch.ValueFrom != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch can not define both value and valueFrom.",
+				))
+		}
+
+		// Attempt to marshal the JSON to discover if it  is valid.
+		if jsonPatch.Value != nil {
+			if err := json.Unmarshal(jsonPatch.Value.Raw, &struct{}{}); err != nil {
+				allErrs = append(allErrs,
+					field.Invalid(
+						path.Index(i),
+						jsonPatch.Value,
+						"jsonPatch Value is invalid JSON.",
+					))
+			}
+		}
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template == nil && jsonPatch.ValueFrom.Variable == nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i).Child("valueFrom"),
+					jsonPatch,
+					"valueFrom must set either \"/template or variable/\".",
+				))
+		}
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template != nil && jsonPatch.ValueFrom.Variable != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i).Child("valueFrom"),
+					jsonPatch,
+					"valueFrom can not set both \"/Template and variable/\".",
+				))
+		}
+
+		// Validate that array access is only prepend or append for add and not allowed for replace.
+		allErrs = append(allErrs,
+			validateIndexAccess(jsonPatch, path.Index(i).Child("path"))...,
+		)
+
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template != nil {
+			// Error if template can not be parsed.
+			_, err := template.New("tpl").Parse(*jsonPatch.ValueFrom.Template)
+			if err != nil {
+				allErrs = append(allErrs,
+					field.Invalid(
+						path.Index(i).Child("valueFrom", "template"),
+						jsonPatch.ValueFrom.Template,
+						fmt.Sprintf("Template can not be parsed. Error %s", err),
+					))
+			}
+		}
+
+		// If set validate that the variable is valid.
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Variable != nil {
+			// If the variable looks like a builtin it's valid.
+			// TODO: Should we try to check if this is a valid builtin?
+			if strings.HasPrefix(*jsonPatch.ValueFrom.Variable, "builtin") {
+				if _, ok := builtinVariables[*jsonPatch.ValueFrom.Variable]; !ok {
+					allErrs = append(allErrs,
+						field.Invalid(
+							path.Index(i).Child("valueFrom", "variable"),
+							jsonPatch.ValueFrom.Variable,
+							"not a defined builtin variable",
+						))
+				}
+				continue
+			}
+			if _, ok := variableSet[*jsonPatch.ValueFrom.Variable]; !ok {
+				allErrs = append(allErrs,
+					field.Invalid(
+						path.Index(i).Child("valueFrom", "variable"),
+						jsonPatch.ValueFrom.Variable,
+						fmt.Sprintf("Variable with name %s cannot be found under spec.variables.", *jsonPatch.ValueFrom.Variable),
```

Comment:
```suggestion
						fmt.Sprintf("variable with name %s cannot be found", *jsonPatch.ValueFrom.Variable),
```
nit, lower case +

I think I would drop the spec variables as it's not necessary and given that this will probably not stay the only source for variables (but feel free to keep it if you want)